connectors/sources/sharepoint_online.py

```diff
@@ -663,118 +719,259 @@ async def validate_config(self):
+
+            self.update_drive_delta_link(drive_id=site_drive_id, link=page.delta_link())
+
+    async def site_list_items(
+        self, site_id, site_list_id, site_web_url, site_list_name
+    ):
+        async for list_item in self.client.site_list_items(site_id, site_list_id):
+            list_item["_id"] = list_item["id"]
+            list_item["object_type"] = "list_item"
+            list_item["_tempfile_suffix"] = os.path.splitext(
+                list_item.get("FileName", "")
+            )[-1]
+
+            content_type = list_item["contentType"]["name"]
+
+            if content_type in [
+                "Web Template Extensions",
+                "Client Side Component Manifests",
+            ]:  # TODO: make it more flexible. For now I ignore them cause they 404 all the time
+                continue
+
+            if "Attachments" in list_item["fields"]:
+                async for list_item_attachment in self.client.site_list_item_attachments(
+                    site_web_url, site_list_name, list_item["id"]
+                ):
+                    list_item_attachment["_id"] = list_item_attachment["odata.id"]
+                    list_item_attachment["object_type"] = "list_item_attachment"
+                    list_item_attachment["_timestamp"] = list_item[
+                        "lastModifiedDateTime"
+                    ]
+                    attachment_download_func = partial(
+                        self.get_attachment_content, list_item_attachment
+                    )
+                    yield list_item_attachment, attachment_download_func
+
+            yield list_item, None
+
+    async def site_lists(self, site_id):
+        async for site_list in self.client.site_lists(site_id):
+            site_list["_id"] = site_list["id"]
+            site_list["object_type"] = "site_list"
+
+            yield site_list
+
+    async def site_pages(self, url):
+        async for site_page in self.client.site_pages(url):
+            site_page["_id"] = site_page["GUID"]
+            site_page["object_type"] = "site_page"
+
+            for html_field in ["LayoutWebpartsContent", "CanvasContent1"]:
+                if html_field in site_page:
+                    site_page[html_field] = html_to_text(site_page[html_field])
+
+            yield site_page
+
+    def initiate_sync_cursor(self):
+        if not self._sync_cursor:
+            self._sync_cursor = {CURSOR_SITE_DRIVE_KEY: {}}
+
+        return self._sync_cursor
+
+    def update_drive_delta_link(self, drive_id, link):
+        self._sync_cursor[CURSOR_SITE_DRIVE_KEY][drive_id] = link
+
+    def get_drive_delta_link(self, drive_id):
+        return self._sync_cursor.get(CURSOR_SITE_DRIVE_KEY, {}).get(drive_id)
+
+    def drive_item_operation(self, item):
+        if "deleted" in item:
+            return OP_DELETE
+        else:
+            return OP_INDEX
+
+    def download_function(self, drive_item, max_data_age):
```

Comment:
This extraction itself is worth a PR, to make this PR cleaner.

Reply:
Could you please add more context because I'm not sure I understand what you are saying here.

Reply:
I meant, this change is to extract a block of code into a new function `download_function`, and this can be a separate PR, so that this PR is smaller and cleaner to review.